src/index.ts

```diff
@@ -0,0 +1,372 @@
+  248: 'RESPONSE_ERROR_GENERATE_MNEMONIC',
+  247: 'RESPONSE_ERROR_INPUT_TIMEOUT',
+  246: 'RESPONSE_ERROR_NOT_IMPLEMENTED'
+};
+
+const STATE_IDLE = 0;
+const STATE_SENDING = 1;
+const STATE_READING = 2;
+
+const state_symbol = Symbol('state');
+const lock_symbol = Symbol('ready');
+const train_symbol = Symbol('train');
+const watchdog_symbol = Symbol('watchdog');
+const reconnect_symbol = Symbol('reconnect');
+
+const WATCHDOG_TIMEOUT = 5000;
+
+var id = 0;
+
+export interface RyderSerialOptions extends SerialPort.OpenOptions {
+  rejectOnLocked: any;
+  reconnectTime: number,
+  debug: boolean;
+}
+
+type TrainEntry = [
+  string,                 // data
+  (value?: any) => void,  // resolve
+  (error?: any) => void,  // reject
+  boolean,                // isEscapedByte
+  // TODO: does this need to be refactored to type `Buffer`?
+  string                  // output buffer
+]
+
+
+export default class RyderSerial extends Events.EventEmitter {
+  id: number = 0;
+  port: string;
+  options: RyderSerialOptions;
+  // TODO: refactor train into its own encapsulated class
+  [train_symbol]: TrainEntry[]
+  [state_symbol]: number;
+  [lock_symbol]: Array<(value?: any) => void>
+  [watchdog_symbol]: NodeJS.Timeout
+  [reconnect_symbol]: NodeJS.Timeout
+
+  closing: boolean;
+  serial?: SerialPort
+
+  static COMMAND_WAKE = 1;
+  static COMMAND_INFO = 2;
+  static COMMAND_SETUP = 10;
+  static COMMAND_RESTORE_FROM_SEED = 11;
+  static COMMAND_RESTORE_FROM_MNEMONIC = 12;
+  static COMMAND_ERASE = 13;
+  static COMMAND_EXPORT_OWNER_KEY = 18;
+  static COMMAND_EXPORT_OWNER_KEY_PRIVATE_KEY = 19;
+  static COMMAND_EXPORT_APP_KEY = 20;
+  static COMMAND_EXPORT_APP_KEY_PRIVATE_KEY = 21;
+  static COMMAND_EXPORT_OWNER_APP_KEY_PRIVATE_KEY = 23;
+  static COMMAND_EXPORT_PUBLIC_IDENTITIES = 30;
+  static COMMAND_EXPORT_PUBLIC_IDENTITY = 31;
+
+  static COMMAND_START_ENCRYPT = 40;
+  static COMMAND_START_DECRYPT = 41;
+
+  static COMMAND_CANCEL = 100;
+
+  static RESPONSE_OK = RESPONSE_OK;
+  static RESPONSE_SEND_INPUT = RESPONSE_SEND_INPUT;
+  static RESPONSE_REJECTED = RESPONSE_REJECTED;
+  static RESPONSE_LOCKED = RESPONSE_LOCKED;
+
+  constructor(port: string, options?: RyderSerialOptions) {
+    super()
+    this.id = id++;
+    this.port = port;
+    this.options = { rejectOnLocked: false, reconnectTime: 1000, debug: false, ...options }
+    this[train_symbol] = [];
+    this[state_symbol] = STATE_IDLE;
+    this[lock_symbol] = [];
+    this.closing = false;
+    this.open(this.port, this.options);
+  }
+
+  serial_error(error: Error) {
+    this.emit('error', error);
+    if (this[train_symbol][0]) {
+      var [, , reject] = this[train_symbol].shift()!;
+      reject(error);
+    }
+    clearTimeout(this[watchdog_symbol]);
+    this[state_symbol] = STATE_IDLE;
+    this.next();
+  }
+
+  serial_data(data: Uint8Array): boolean | void {
+    this.options.debug && console.debug('data from Ryder', '0x' + data.toString());
```

Comment:
this used to feature `data.toString('hex')` which TypeScript ruled out. Should the `data` param be something other than type `Uint8Array`?

Reply:
fixed in 86fc02f.